src/Microsoft.DotNet.Darc/src/DarcLib/Actions/Remote.cs

```diff
@@ -281,41 +281,320 @@ public async Task MergePullRequestAsync(string pullRequestUrl, MergePullRequestP
+                            nodeCache.Add($"{node.RepoUri}@{node.Commit}", node);
+                        }
+                    }
+                    rootNode = dependencyGraph.Root;
+                }
+
+                // Now do the lookup to find the element in the tree for each item in the update list
+                foreach (DependencyDetail dependencyInUpdateChain in updateList)
+                {
+                    (Asset coherentAsset, Build buildForAsset) =
+                        FindAssetInBuildTree(dependencyInUpdateChain.Name, rootNode);
+
+                    if (coherentAsset == null)
+                    {
+                        // This is an invalid state. We can't satisfy the
+                        // constraints so they should either be removed or pinned.
+                        throw new DarcException($"Unable to update {dependencyInUpdateChain.Name} to have coherency with " +
+                            $"parent {dependencyInUpdateChain.CoherentParentDependencyName}. No matching asset found in tree. " +
+                            $"Either remove the coherency attribute or mark as pinned.");
+                    }
+
+                    string buildRepoUri = buildForAsset.AzureDevOpsRepository ?? buildForAsset.GitHubRepository;
 
-        public async Task<List<DependencyDetail>> GetRequiredUpdatesAsync(
+                    if (dependencyInUpdateChain.Name == coherentAsset.Name &&
+                        dependencyInUpdateChain.Version == coherentAsset.Version &&
+                        dependencyInUpdateChain.Commit == buildForAsset.Commit &&
+                        dependencyInUpdateChain.RepoUri == buildRepoUri)
+                    {
+                        continue;
+                    }
+
+                    DependencyDetail updatedDependency = new DependencyDetail(dependencyInUpdateChain);
+                    updatedDependency.Name = coherentAsset.Name;
+                    updatedDependency.Version = coherentAsset.Version;
+                    updatedDependency.RepoUri = buildForAsset.AzureDevOpsRepository ?? buildForAsset.GitHubRepository;
+                    updatedDependency.Commit = buildForAsset.Commit;
+
+                    toUpdate.Add(dependencyInUpdateChain, updatedDependency);
+                    visited.Add(dependencyInUpdateChain);
+                }
+            }
+
+            return toUpdate;
+        }
+
+        /// <summary>
+        ///     Given an asset name, find the asset in the dependency tree.
+        ///     Returns the asset with the shortest path to the root node.
+        /// </summary>
+        /// <param name="assetName">Name of asset.</param>
+        /// <param name="currentNode">Dependency graph node to find the asset in.</param>
+        /// <returns>(Asset, Build, depth), or (null, null, maxint) if not found.</returns>
+        private (Asset, Build, int) FindAssetInBuildTree(string assetName, DependencyGraphNode currentNode, int currentDepth)
```

Comment:
>Asset [](start = 17, length = 5)

We should probably name these 3 tuple chunks... Especially since the last one is just a random int.